taglibs/core/TagHandlerNode.js

```diff
@@ -177,17 +178,42 @@ TagHandlerNode.prototype = {
         template.functionCall(tagHelperVar, function () {
             template.code('out,\n').indent(function () {
                 template.line(handlerVar + ',').indent();
+
+                if (_this.dynamicAttributes) {
+                    template.indent(function() {
+                        _this.setProperty(_this.dynamicAttributesProperty, template.makeExpression(getPropsStr(_this.dynamicAttributes, template)));
+                    });
+                }
+
+                var props = _this.getProperties();
+
+                var propsCode = getPropsStr(props, template);
+
                 if (_this.inputExpression) {
-                    template.code(_this.inputExpression);
-                } else {
-                    if (_this.dynamicAttributes) {
-                        template.indent(function() {
-                            _this.setProperty(_this.dynamicAttributesProperty, template.makeExpression(getPropsStr(_this.dynamicAttributes, template)));
-                        });
+                    if (isObjectEmpty(props)) {
+                        propsCode = _this.inputExpression;
+                    } else {
+                        // We need to generate code that merges in the attribute properties with
+                        // the provided data object. We don't want to modify the existing
+                        // data object provided by the user so first need to create a new
+                        // empty object and then merge in the existing properties from the
+                        // provided object. When then extend that object with the properties
+                        // that came from the attributes.
+                        //
+                        // The generated code will be similar to the following:
+                        //
+                        // extend(extend({}, <input_expression>), <attr_props>);
+                        var extendVar = template.addStaticVar('__extend', '__helpers.xt');
```

Comment:
The `template.addStaticVar('__extend', '__helpers.xt');` line will result in a new JavaScript variable introduced in the template with the following code:

```javascript
var __extend = __helpers.xt;
```

The value of `extendVar` will be the name of the variable (i.e. `"__extend"`).